### PR TITLE
create slushy_trim

### PR DESCRIPTION
**Pull request by karima662**
This PR is addressing issue #28

### CODE REVIEW

File: R/slushy_trim.R [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  confirm_clean_env()
+  
+  # List of packages currently installed in project library
+  proj_library_pkgs <- rownames(installed.packages(paths$library())) %>% unique()
+  
+  # List of packages used in code files (excluding ones in Rprofile/DESCRIPTION)
+  used_pkgs <- renv::dependencies(root = project, progress = FALSE) %>%
+    as_tibble() %>% 
+    filter(!str_detect(Source, regex("\\.Rprofile$", ignore_case = TRUE)) &
+             !str_detect(Source, regex("DESCRIPTION$", ignore_case = TRUE))) %>%
```
Comment (bzkrouse) [inline R/slushy_trim.R:39]: can you please also include renv.lock as it will show a dependency of renv

File: R/slushy_trim.R [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    pull(Package) %>%
+    unique()
+  
+  # List of dependencies of packages used
+  used_pkgs_deps <- package_dependencies(packages = used_pkgs,
+                                         db = installed.packages(),
+                                         which = c("Depends", "Imports"),
```
Comment (bzkrouse) [inline R/slushy_trim.R:46]: I believe we also want to include "LinkingTo" here

File: R/slushy_trim.R [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                                         recursive = FALSE) %>%
+    unlist() %>%
+    unique()
+  
+  # Keep packages that are used and their dependencies 
+  pkgs_to_keep <- append(used_pkgs, used_pkgs_deps) %>% unique()
+  
+  # Drop packages not in the keep list
+  pkgs_to_drop <- setdiff(proj_library_pkgs, pkgs_to_keep)
```
Comment (bzkrouse) [inline R/slushy_trim.R:55]: Can you figure out which of these are top-level packages (i.e. listed in DESCRIPTION)? We don't need to notify user that dependencies are being dropped, just top level packages.

File: R/slushy_trim.R [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  
+  # Notify user about packages to be trimmed
+  if (length(pkgs_to_drop) > 0) {
+    cli_alert_info(paste("The following packages are not used and will be removed:\n\n",
+                         paste(pkgs_to_drop, collapse = ", "), "\n\n"))
+    
+    # Ask user if they want to proceed
+    cat("Would you like to proceed? \n1. Proceed \n2. Cancel \nSelection: ")
```
Comment (bzkrouse) [inline R/slushy_trim.R:63]: can you try utils::menu here instead? I think it will be simpler!

File: R/slushy_trim.R [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    # Flush output to make sure prompt is shown before reading input
+    flush.console()
+    
+    # Wait for user input
+    choice <- readline()
+    
+    ## Choice 1
+    if (choice == "1") {
+      # Drop unused packages
+      slushy_drop(pkgs_to_drop)
```
Comment (bzkrouse) [inline R/slushy_trim.R:74]: if you subset this to just the top-level packages above, we will need to check this effectively removes all deps we dont need anymore.

File: R/slushy_trim.R [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    ## Choice 2
+    } else if (choice == "2") {
+      # Notify user of operation cancel
+      stop("Trimming operation cancelled.")
+    }
+    
+  } else {
+    cli_alert_info("No packages need to be trimmed.")
+    
+    return(invisible(NULL))
```
Comment (bzkrouse) [inline R/slushy_trim.R:88]:
```suggestion
```

this is covered by the last invisible(NULL) right?

File: R/slushy_trim.R (hunk 5 above)
Comment (bzkrouse) [inline R/slushy_trim.R:74]:
```suggestion
      slushy_drop(pkgs_to_drop, restart = FALSE)
```

**Review by bzkrouse — CHANGES_REQUESTED** [review]
this is looking so great!! Just a couple of changes needed!

File: R/slushy_trim.R (hunk 1 above)
Comment (karima662) [reply R/slushy_trim.R]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/cdce33bbefe614d3bfa07ad9825d66ba013e8538

File: R/slushy_trim.R (hunk 2 above)
Comment (karima662) [reply R/slushy_trim.R]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/c002c5e19503aa7c550c835f2e186a17d23878a4

File: R/slushy_trim.R (hunk 4 above)
Comment (karima662) [reply R/slushy_trim.R]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/7aca19fe05bc3f1c950d26ed56a3224883a1c82c

File: R/slushy_trim.R (hunk 6 above)
Comment (karima662) [reply R/slushy_trim.R]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/4c887d9f13ab0dc580c380ef8735d6ddf73bd524

File: R/slushy_trim.R (hunk 5 above)
Comment (karima662) [reply R/slushy_trim.R]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/e43a8840659085a29d8e91741cb9db1e7f16aa71

File: R/slushy_trim.R (hunk 3 above)
Comment (karima662) [reply R/slushy_trim.R:61]: https://github.com/GSK-Biostatistics/slushy/pull/37/commits/3ed9ec9154b75aa07feb7bb2006d987d5cc947c5

created a new list `desc_pkgs` of description packages and `pkgs_to_drop_top_level` which identifies the pkgs in common between `desc_pkgs` and `pkgs_to_drop`. The `pkgs_to_drop_top_level` are output in the message to the user but the `pkgs_to_drop` (top-level plus deps) are the ones being removed in the background.

Comment (karima662) [reply R/slushy_trim.R:61]: Happy to make changes if this isn't what you had in mind

File: R/slushy_trim.R (hunk 5 above)
Comment (karima662) [reply R/slushy_trim.R]: Please see comment in https://github.com/GSK-Biostatistics/slushy/pull/37#discussion_r1761479991

**Review by bzkrouse — APPROVED** [review]
this looks really great, thank you!!!

**Pull request merged**